test/ecto/repo_test.exs

```diff
@@ -378,4 +381,60 @@ defmodule Ecto.RepoTest do
       assert TestRepo.preload(nil, []) == nil
     end
   end
+
+  test "Repo.load/2" do
+    # string fields
+    assert %MySchema{x: "abc"} =
+           TestRepo.load(MySchema, %{"x" => "abc"})
+
+    # atom fields
+    assert %MySchema{x: "abc"} =
+           TestRepo.load(MySchema, %{x: "abc"})
+
+    # keyword list
+    assert %MySchema{x: "abc"} =
+           TestRepo.load(MySchema, [x: "abc"])
+
+    # atom fields and values
+    assert %MySchema{x: "abc"} =
+           TestRepo.load(MySchema, {[:x], ["abc"]})
+
+    # string fields and values
+    assert %MySchema{x: "abc"} =
+           TestRepo.load(MySchema, {["x"], ["abc"]})
+
+    # default value
+    assert %MySchema{x: "abc", z: "z"} =
+           TestRepo.load(MySchema, %{x: "abc"})
+
+    # array field
+    assert %MySchema{array: ["one", "two"]} =
+           TestRepo.load(MySchema, %{array: ["one", "two"]})
+
+    # map field with atoms
+    assert %MySchema{map: %{color: "red"}} =
+           TestRepo.load(MySchema, %{map: %{color: "red"}})
+
+    # map field with strings
+    assert %MySchema{map: %{"color" => "red"}} =
+           TestRepo.load(MySchema, %{map: %{"color" => "red"}})
+
+    # nil
+    assert %MySchema{x: nil} =
+           TestRepo.load(MySchema, %{x: nil})
+
+    # invalid field is ignored
+    assert %MySchema{} =
+           TestRepo.load(MySchema, %{bad: "bad"})
+
+    # FIXME
+    # # invalid value
+    # assert_raise ArgumentError, "cannot load `0` as type :string for :x in schema Ecto.RepoTest.MySchema", fn ->
+    #   TestRepo.load(MySchema, %{x: 0})
+    # end
```

Comment:
The difference is that adapters may customize some types. For example MySQL will be loading `0`, `1`, `<<0>>`, `<<1>>` as booleans (since it lacks native boolean types). There's also special handling for json - mysql will load json-encoded strings and all SQL adapters will have special handling for loading embeds. Furthermore, every adapter must specify how to load the special `binary_id` type that is adapter-dependent.

The lack of error here is strange... It should be wrong.

@josevalim Is that because of this bypass? https://github.com/elixir-ecto/ecto/blob/master/lib/ecto/type.ex#L631